test/infrastructure/docker/internal/provisioning/cloudinit/writefiles.go

```diff
@@ -62,13 +66,79 @@ func newWriteFilesAction() action {
 	return &writeFilesAction{}
 }
 
-func (a *writeFilesAction) Unmarshal(userData []byte) error {
+func (a *writeFilesAction) Unmarshal(userData []byte, kindMapping kind.Mapping) error {
 	if err := yaml.Unmarshal(userData, a); err != nil {
 		return errors.Wrapf(err, "error parsing write_files action: %s", userData)
 	}
+	for i, f := range a.Files {
+		if f.Path == kubeadmInitPath {
+			// NOTE: in case of join the kubeadmConfigFile contains both the ClusterConfiguration and the InitConfiguration
+			contentSplit := strings.Split(f.Content, "---\n")
+
+			if len(contentSplit) != 3 {
+				return errors.Errorf("invalid kubeadm config file, unable to parse it")
+			}
+			initConfiguration, err := kubeadmtypes.UnmarshalInitConfiguration(contentSplit[2])
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse init configuration")
+			}
+
+			fixNodeRegistration(&initConfiguration.NodeRegistration, kindMapping.Mode)
+
+			contentSplit[2], err = kubeadmtypes.MarshalInitConfigurationForVersion(initConfiguration, kindMapping.KubernetesVersion)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse init configuration")
+			}
+			a.Files[i].Content = strings.Join(contentSplit, "---\n")
+		}
+		if f.Path == kubeadmJoinPath {
+			// NOTE: in case of join the kubeadmConfigFile contains only the join Configuration
+			joinConfiguration, err := kubeadmtypes.UnmarshalJoinConfiguration(f.Content)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse join configuration")
+			}
+
+			fixNodeRegistration(&joinConfiguration.NodeRegistration, kindMapping.Mode)
+
+			a.Files[i].Content, err = kubeadmtypes.MarshalJoinConfigurationForVersion(joinConfiguration, kindMapping.KubernetesVersion)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse join configuration")
+			}
+		}
+	}
 	return nil
 }
 
+// fixNodeRegistration sets node registration for running Kubernetes/kubelet in docker.
+// NOTE: we add those values if they do not exists; user can set those flags to different values to disable automatic fixing.
+// NOTE: if there will be use case for it, we might investigate better ways to disable automatic fixing.
+func fixNodeRegistration(nodeRegistration *bootstrapv1.NodeRegistrationOptions, kindMode kind.Mode) {
+	if nodeRegistration.CRISocket == "" {
+		// TODO: check if we can drop this, kubeadm CRI auto detection should be enough.
```

Comment:
I think it is not enough. The reason is that for self-hosted cluster we have to mount the Docker socket. On those nodes we have the Docker and the containerd socket and then kubelet wouldn't know which one to use.

I think we can drop the todo and document this instead?